src/Raven.Server/Documents/Patch/V8/ScriptRunner.cs

```diff
@@ -0,0 +1,1700 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using V8.Net;
+using Raven.Server.Extensions.V8;
+using Raven.Client;
+using Raven.Client.Documents.Indexes.Spatial;
+using Raven.Client.Documents.Session.TimeSeries;
+using Raven.Client.Exceptions.Documents;
+using Raven.Client.Exceptions.Documents.Patching;
+using Raven.Server.ServerWide;
+using Raven.Server.Documents.Indexes.Static.Spatial;
+using Raven.Server.Documents.Queries.Results.TimeSeries;
+using Raven.Server.Documents.Queries.Timings;
+using Raven.Server.Documents.TimeSeries;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow;
+using Sparrow.Extensions;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Spatial4n.Core.Distance;
+using ExpressionType = System.Linq.Expressions.ExpressionType;
+using Raven.Server.Documents.Patch.V8;
+
+namespace Raven.Server.Documents.Patch
+{
+    public partial class ScriptRunner
+    {
+        public static unsafe DateTime GetDateArg(InternalHandle arg, string signature, string argName)
+        {
+            if (arg.IsDate)
+                return arg.AsDate;
+
+            if (arg.IsStringEx == false)
+                ThrowInvalidDateArgument();
+
+            var s = arg.AsString;
+            fixed (char* pValue = s)
+            {
+                var result = LazyStringParser.TryParseDateTime(pValue, s.Length, out DateTime dt, out _);
+                if (result != LazyStringParser.Result.DateTime)
+                    ThrowInvalidDateArgument();
+
+                return dt;
+            }
+
+            void ThrowInvalidDateArgument() =>
+                throw new ArgumentException($"{signature} : {argName} must be of type 'DateInstance' or a DateTime string. {GetTypes(arg)}");
+        }
+
+        private static DateTime GetTimeSeriesDateArg(InternalHandle arg, string signature, string argName)
+        {
+            if (arg.IsDate)
+                return arg.AsDate;
+
+            if (arg.IsStringEx == false)
+                throw new ArgumentException($"{signature} : {argName} must be of type 'DateInstance' or a DateTime string. {GetTypes(arg)}");
+
+            return TimeSeriesRetriever.ParseDateTime(arg.AsString);
+        }
+        
+        private static string GetTypes(InternalHandle value) => $"JintType({value.ValueType}) .NETType({value.GetType().Name})";
+        
+        public partial class SingleRun
+        {
+            public V8EngineEx ScriptEngineExV8;
+            public V8Engine ScriptEngineV8;
```

Comment:
why both fields?

Reply:
in V8EngineEx we have overloaded methods that return generic JsHandle instead of V8.Net's InternalHandle (like CreateValue methods)
in general we operate on the generic level (V8EngineEx), but in callbacks we operate on the V8.Net level for speed (V8Engine)